### DISCUSSION TITLE
CancellationToken Support (added in 1.4.10)

### DISCUSSION THREAD

**Discussion by berkayk** [KyryloKuzyk/PrimeTween#46] [FAQ] [answered]
Hi, 
I couldn't find a way to integrate cancellation tokens. How can I use them?
Thanks.

**Comment by KyryloKuzyk** [comment, marked as answer]
This was a design decision not to rely on System.Threading namespace. PrimeTween supports async/await simply by adhering to C# language [specification](https://www.jacksondunstan.com/articles/4918). This ensures that PrimeTween can work on platforms that don't support threads, like WebGL.

So I'm afraid that there is no support for CancellationToken. The closest alternative is to use [tween/sequence.Stop()](https://github.com/KyryloKuzyk/PrimeTween?tab=readme-ov-file#controlling-tweens) if you need to interrupt an animation midway.

**Reply by NOMOAX** [reply]
> This was a design decision not to rely on System.Threading namespace. PrimeTween supports async/await simply by adhering to C# language [specification](https://www.jacksondunstan.com/articles/4918). This ensures that PrimeTween can work on platforms that don't support threads, like WebGL.
> 
> So I'm afraid that there is no support for CancellationToken. The closest alternative is to use [tween/sequence.Stop()](https://github.com/KyryloKuzyk/PrimeTween?tab=readme-ov-file#controlling-tweens) if you need to interrupt an animation midway.

The `CancellationToken` can be used independently and does not necessarily have to be used in conjunction with `Thread` or `Task`. I think it is reasonable to add support for `CancellationToken`.

**Reply by KyryloKuzyk** [reply]
@Nomoax Yes, I agree. I'm already considering adding CancellationToken support.

**Comment by berkayk** [comment]
Thanks.
Sometimes the player leaves the app in the background, and animation resumes when they return. This is a multiplayer game, so these differences create some problems in the game state such as one player resumes the game while the other still watches the animation. How can I solve this? Maybe I can listen to some `AnimationStopped` action and close the relevant tweens(sequences).

**Reply by KyryloKuzyk** [reply]
@berkayk I don't have a full context to give an exact answer for your use case, but here are a few options from the top of my mind:
1. Set `useUnscaledTime = true`, so animations are not tied to the ingame time, but to the real time. This way after exiting the background animations will jump forward to accommodate for the passed time.
2. If you have your own mechanism that synchronizes time between the server and peers, you can use the `elapsedTime` property to manually set the elapsed time of tweens and sequences.

**Reply by berkayk** [reply]
Wish I knew `useUnscaledTime` before :/ That's why you have to read the f manual 😅 
Thanks!

**Comment by KyryloKuzyk** [comment]
Hey @berkayk @Nomoax, I added CancellationToken support in version 1.4.10.

**Reply by JoanOrtiga-AntiHero** [reply]
Im in favour of the addition but, can I ask what made you change your mind and add it? Thanks.

**Reply by KyryloKuzyk** [reply]
@JoanOrtiga-AntiHero The sheer number of requests. 

Also, the curiosity of learning the intricacies of how CancellationToken works. It was very fun because I learned 100 ways to shoot yourself in the foot with them 😀
